bundle/edu.gemini.qpt.client/src/main/java/edu/gemini/qpt/core/listeners/LimitsListener.java

```diff
@@ -238,3 +243,51 @@ protected MarkerManager getMarkerManager(Variant t) {
 	}
 
 }
+
+
+// REL-397: warn if another unscheduled obs in the same program has a higher score and is within 1.5h in RA.
+class ScoreMarker {
+
+	private static final double RA_LIMIT_H = 1.5; // we only care about other observations within this distance in RA
+	private static final double RA_LIMIT_DEG = RA_LIMIT_H * 15; // 15 degrees per hour
+	private static final String PREFIX = String.format("Higher-scoring observation(s) within %2.1fH: ", RA_LIMIT_H);
+
+	public static void add(Object owner, MarkerManager mm, Variant v, Alloc a) {
+
+		// unscheduled observations
+		final Predicate<Obs> unscheduled = obs ->
+			!v.getFlags(obs).contains(Flag.SCHEDULED);
+
+		// observations within RA_LIMIT_DEG of `a`
+		final Predicate<Obs> nearby = otherObs -> {
+			final double myRaDeg    = a.getObs().getRa(a.getStart()); // My RA at scheduling time
+			final double otherRaDeg = otherObs.getRa(a.getStart());
+			final double raDiffDeg  = Math.abs(myRaDeg - otherRaDeg);
+			return raDiffDeg <= RA_LIMIT_DEG;
+		};
+
+		// observations with a better score than `a.getObs`
+		final Predicate<Obs> higherScoring = otherObs -> {
+			final double myScore    = v.getScore(a.getObs());
+			final double otherScore = v.getScore(otherObs);
+			return otherScore > myScore;
+		};
+
+		// all of the above, ordered by cost to compute
+		final Predicate<Obs> all =
+			unscheduled.and(nearby).and(higherScoring);
```

Comment:
I did not know about `Predicate.and` and `Predicate.or`. Neat. (Well, by Java standards.)